src/index/am/mod.rs

```diff
@@ -557,17 +557,50 @@ impl Drop for HeapFetcher {
 }
 
 impl SearchFetcher for HeapFetcher {
+    /// Fetches data associated with the given `key` passes the filter criteria.
+    ///
+    /// # Parameters
+    /// - `key`: A 3-element array representing the key to be checked.
+    ///
+    /// Returns a tuple containing a reference to an array of `Datum` and a reference to an array of `bool`,
+    /// or `None` if no data is found.
     fn fetch(&mut self, key: [u16; 3]) -> Option<(&[Datum; 32], &[bool; 32])> {
+        if !self.filter(key) {
+            return None;
+        }
+        unsafe {
+            (*self.econtext).ecxt_scantuple = self.slot;
+            pgrx::pg_sys::MemoryContextReset((*self.econtext).ecxt_per_tuple_memory);
+            pgrx::pg_sys::FormIndexDatum(
+                self.index_info,
+                self.slot,
+                self.estate,
+                self.values.as_mut_ptr(),
+                self.is_nulls.as_mut_ptr(),
+            );
+            Some((&self.values, &self.is_nulls))
+        }
+    }
+
+    /// Determines whether the given `key` passes the filter criteria.
+    ///
+    /// # Parameters
+    /// - `key`: A 3-element array representing the key to be checked.
+    ///
+    /// # Returns
+    /// - `true` if the key satisfies the filter conditions.
+    /// - `false` otherwise.
+    fn filter(&self, key: [u16; 3]) -> bool {
         unsafe {
             let mut ctid = key_to_ctid(key);
             let table_am = (*self.heap_relation).rd_tableam;
             let fetch_row_version = (*table_am)
                 .tuple_fetch_row_version
                 .expect("unsupported heap access method");
             if !fetch_row_version(self.heap_relation, &mut ctid, self.snapshot, self.slot) {
-                return None;
+                return false;
             }
-            if !self.hack.is_null() && prererank_filtering() {
+            if !self.hack.is_null() && enable_prefilter() {
```

Comment:
It should not be checked here. It should be checked eariler. If `prefilter` is `false` and `rerank_in_heap` is `false`, `HeapFetcher` should leave untouched.